cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -189,6 +191,72 @@ Status ConstBoolExec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
   return Status::OK();
 }
 
+struct NonZeroVisitor {
+   UInt32Builder *builder;
+   const ArrayData& array;
+
+   NonZeroVisitor(UInt32Builder *builder, const ArrayData& array)
+   : builder(builder), array(array) {}
+
+  Status Visit(const DataType& type) {
+    return Status::TypeError("Unsupported type for nonzero: ", type.ToString());
+  }
+
+   template <typename Type> 
+   enable_if_t<has_c_type<Type>::value && 
+              !std::is_same<Type, MonthDayNanoIntervalType>::value &&
+              !std::is_same<Type, DayTimeIntervalType>::value, 
+              Status>
+   Visit(const Type&) {
+     using T = typename GetViewType<Type>::T;
+     uint32_t index = 0;
+
+     return VisitArrayDataInline<Type>(
+        this->array,
+        [&](T v) {
+          if(v != 0) {
+            RETURN_NOT_OK(this->builder->Reserve(1));
+            this->builder->UnsafeAppend(index);
```

Comment:
[Use `builder->Append(index)`](https://github.com/edponce/arrow/blob/master/cpp/src/arrow/array/builder_primitive.h#L88)

Reply:
@edponce that would be quite slow, isn't it? I think builder should reserve initially to array length and then do unsafeappend

Reply:
Agree. You can reserve builder with same size as input array and then do a `Resize` at the end. Thanks @nirandaperera!

Reply:
Fair one, it was actually the original implementation but I was getting a 
```
  input.array()->length
    Which is: 4
  expected->length()
    Which is: 2
```
error which I thought it was due to the fact that I was reserving 4 entries even thought I was only willing to write 2. But in the end I found that it was due to the `CheckScalarUnary` helper and got rid of it. Thus I can now put back the single `Reserve` call that I had before

Reply:
> 
I believe `builder->Finalize(...)` does the resizing internally.

Reply:
@edponce I believe `builder->Finalize(...)` does the resizing internally.

Reply:
`builder->Finalize(...)` performs a resizing of sorts. It does not calls `Resize` but it creates an `ArrayData` of length equal to the number of used values in `builder`, so essentially a resize behavior.

Reply:
Sorry it's `builder->Finalize()` and yes, it does the shrinking internally.

Reply:
Well, `data_builder_.FinishWithLength(length_)` is actually, `data_builder_.FinishWithLength(length_, /*shrink_to_fit=*/ true)`. So, it should resize the buffers as well.

Reply:
put back the single `Resize` when the builder is created